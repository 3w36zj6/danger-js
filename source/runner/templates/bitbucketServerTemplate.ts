import { DangerResults } from "../../dsl/DangerResults"
import { Violation } from "../../dsl/Violation"

/**
 * Converts a set of violations into a Markdown section
 *
 * @param {string} name User facing title of section
 * @param {string} emoji Emoji name to show
 * @param {Violation[]} violations for section
 * @returns {string} Markdown
 */
function resultsSection(name: string, emoji: string, violations: Violation[]): string {
  if (violations.length === 0 || violations.every(violation => !violation.message)) {
    return ""
  }
  return (
    `\n` +
    `| ${emoji} | ${name}\n` +
    `| --- | --- |\n` +
    `\n` +
    violations
      .map(v => {
        return (
          "> " +
          v.message
            .replace(/<\/?code>/g, "`")
            .split("\n")
            .join("\n> ")
        )
      })
      .join("\n\n") +
    `\n`
  )
}

export const dangerIDToString = (id: string) => `danger-id-${id};`

/**
 * Postfix signature to be attached comment generated / updated by danger.
 */
export const dangerSignaturePostfix = `
|    |
|---:|
| _Generated by 🚫 [dangerJS](http://github.com/danger/danger-js/)_ |
`

/**
 * A template function for creating a GitHub issue comment from Danger Results
 * @param {string} dangerID A string that represents a unique build
 * @param {DangerResults} results Data to work with
 * @returns {string} HTML
 */
export function template(dangerID: string, results: DangerResults): string {
  return `
${resultsSection("Fails", "🚫", results.fails)}
${resultsSection("Warnings", "⚠️", results.warnings)}
${resultsSection("Messages", "📖", results.messages)}

${results.markdowns.join("\n\n")}

${dangerSignaturePostfix}

[](http://${dangerIDToString(dangerID)})
`
}
