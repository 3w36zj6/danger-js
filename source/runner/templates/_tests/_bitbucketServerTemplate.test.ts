import {
  emptyResults,
  failsResultsWithoutMessages,
  warnResults,
  failsResults,
  summaryResults,
  messagesResults,
  markdownResults,
} from "../../_tests/fixtures/ExampleDangerResults"
import { template, inlineTemplate } from "../bitbucketServerTemplate"

const noEntryEmoji = "\u274C"
const warningEmoji = "⚠️"
const messageEmoji = "\u2728"

describe("generating messages for BitBucket server", () => {
  it("shows no sections for empty results", () => {
    const issues = template("blankID", emptyResults)
    expect(issues).not.toContain("Fails")
    expect(issues).not.toContain("Warnings")
    expect(issues).not.toContain("Messages")
  })

  it("shows no sections for results without messages", () => {
    const issues = template("blankID", failsResultsWithoutMessages)
    expect(issues).not.toContain("Fails")
    expect(issues).not.toContain("Warnings")
    expect(issues).not.toContain("Messages")
  })

  it("Shows the failing messages in a section", () => {
    const issues = template("blankID", failsResults)
    expect(issues).toContain("Fails")
    expect(issues).not.toContain("Warnings")
  })

  it("Shows the warning messages in a section", () => {
    const issues = template("blankID", warnResults)
    expect(issues).toContain("Warnings")
    expect(issues).not.toContain("Fails")
  })

  it("summary result matches snapshot", () => {
    expect(template("blankID", summaryResults)).toMatchSnapshot()
  })
})

describe("generating inline messages", () => {
  it("Shows the failing message", () => {
    const issues = inlineTemplate("blankID", failsResults, "File.swift", 5)
    expect(issues).toContain("- ${noEntryEmoji} Failing message")
    expect(issues).not.toContain("- ${warningEmoji}")
    expect(issues).not.toContain("- ${messageEmoji}")
  })

  it("Shows the warning message", () => {
    const issues = inlineTemplate("blankID", warnResults, "File.swift", 5)
    expect(issues).toContain("- ${warningEmoji} Warning message")
    expect(issues).not.toContain("- ${noEntryEmoji}")
    expect(issues).not.toContain("- ${messageEmoji}")
  })

  it("Shows the message", () => {
    const issues = inlineTemplate("blankID", messagesResults, "File.swift", 5)
    expect(issues).toContain("- ${messageEmoji} Message")
    expect(issues).not.toContain("- ${noEntryEmoji}")
    expect(issues).not.toContain("- ${warningEmoji}")
  })

  it("Shows markdowns one after another", () => {
    const issues = inlineTemplate("blankID", markdownResults, "File.swift", 5)
    const expected = `
### Short Markdown Message1

### Short Markdown Message2
`
    expect(issues).toContain(expected)
  })
})
