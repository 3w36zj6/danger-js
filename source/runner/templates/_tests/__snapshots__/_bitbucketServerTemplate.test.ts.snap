// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generating messages for BitBucket server summary result matches snapshot 1`] = `
"

| ❌ | Fails
| --- | --- |

> Failing message Failing message


| ⚠️ | Warnings
| --- | --- |

> Warning message Warning message


| ✨ | Messages
| --- | --- |

> message


markdown


|    |
|---:|
| _Generated by ❌ [dangerJS](http://github.com/danger/danger-js/)_ |


[](http://danger-id-blankID;)
"
`;
